### DISCUSSION TITLE
Feeding Maven cache with p2 repo, to speed up tycho build

### DISCUSSION THREAD

**Discussion by fbergomi** [eclipse-tycho/tycho#755] [Q&A]
Hello all,
I'm happily using tycho on my continuous integration, and since the build occurs on an empty docker image, the build spends 98% of its time downloading parts of my target platform, plus dependencies, 1% to download tycho and maven-specific artifacts, and 1% to do the build itself.

I would love so much to spare the bandwidth and time spent on the first 98% of my builds!! 

I thought about keeping the maven cache between the builds but I found it difficult, even with precise cleaning actions, since it would create an unwanted "cache effect" that would become a pain for building products using diverse and colliding target platforms (some more recent, some older). 

Si since my target platform varies across the products I am building, I was wondering if it is possible to "mavenize" a clearly identified p2 repository and store it into a maven cache that I would put into each of my "build" docker images.
This would prevent maven to download each time the adhoc target platform (a p2-ized eclipse).

Is this possible in some way with tycho? 
If not, does any one have some idea experience or advice? 

Thank you all! Especially to tycho developers who do fantastic work!
Franco

**Comment by laeubi** [comment]
> I thought about keeping the maven cache between the builds but I found it difficult, even with precise cleaning actions, since it would create an unwanted "cache effect" that would become a pain for building products using diverse and colliding target platforms (some more recent, some older).

Could you explain this? We might be able to fix this instead.

> Si since my target platform varies across the products I am building, I was wondering if it is possible to "mavenize" a clearly identified p2 repository and store it into a maven cache

You could use a reference to a zipped repository with maven urls see https://github.com/eclipse/tycho/blob/master/RELEASE_NOTES.md#support-for-consuming-maven-artifacts-made-of-zipped-p2-update-sites-

> f not, does any one have some idea experience or advice?

In general I always cache the m2 repository without any problem, if you have problems with distinct *builds* it might help to assign a custom repository per build (eg. ~/.m2/repository-build1, ~/.m2/repository-build2, ~/.m2/repository-build...,.)

One key might be to disable local build artifacts with `-Dtycho.localArtifacts=ignore` but then your build needs to be self-contained

**Reply by fbergomi** [reply]
Hi, I finished by using yoru solution : we feed the the m2 repository and store it in our docker image.
I obtain the behaviour by building a dummy plug-in dependending on each plugin contained in my target platform.
This way, the docker image contains a pre-filled .m2 and the build runs without downloading any dependency (except tycho-specific stuff).

I'm trying now to have also features in my .m2, since I need to build some feature-patches.
For the moment I have set up the same kind of process with a dummy plug-in and this time a dummy-feature.
I need to figure out why some features seem to miss in the resulting .m2 cache. 
=> will keep informed as soon as I have some better information
